test/unit/communication/test_transform_mpi.cpp

```diff
@@ -0,0 +1,131 @@
+};
+
+TEST_F(TransformMPITest, PromiseGuardManagement) {
+  static_assert(NUM_MPI_RANKS >= 2, "This test requires 2 ranks");
+
+  // Note:
+  // Rank 0 is where the test actually is done, while Rank 1 is just used by Rank 0 to control
+  // its own progress.
+  //
+  // Rank 0 posts an IRecv that will be satisfied by Rank 1 just after it receives from Rank 0
+  // the trigger. So, actually, Rank 0 can control when to unlock its IRecv by deciding when to
+  // send the trigger to Rank 1. In this way, Rank 0 flow can be "paused" for verifiying various
+  // phases of the async communication mechanism for IRecv.
+
+  Communicator world(MPI_COMM_WORLD);
+  const IndexT_MPI rank = world.rank();
+
+  if (rank == 0) {
+    using dlaf::comm::internal::transformMPI;
+    using dlaf::internal::whenAllLift;
+
+    namespace ex = pika::execution::experimental;
+
+    common::Pipeline<Communicator> chain(world);
+
+    // Note:
+    // `sent_guard` represents the status of completion of IRecv. It will be set true by a `then`
+    // task that depends on completion of transformMPI, i.e. it will become true only when IRecv
+    // communication completes (i.e. posted + message received).
+    std::atomic_bool sent_guard = false;
+
+    int message;
+    whenAllLift(&message, 1, MPI_INT, 1, 0, chain()) | transformMPI(MPI_Irecv) |
+        ex::then([&sent_guard](auto mpi_err_code) {
+          EXPECT_EQ(MPI_SUCCESS, mpi_err_code);
+          sent_guard = true;
+        }) |
+        ex::ensure_started();
+
+    // Note:
+    // At this point IRecv is (getting) posted but it won't complete until this Rank 0 will trigger
+    // Rank 1 to send the message. So, here we can check that PromiseGuard<Communicator> gets
+    // consumed just after MPI operation is posted.
+    //
+    // For checking it gets released, we ask the Pipeline for the next PromiseGuard<Communicator> in
+    // the chain. Indeed, if the previous one is released, this one will be unlocked.
+    //
+    // Let's use a guard + try_waiting_guard, so that in case something goes wrong it does not end
+    // up creating a deadlock. The assumption is that try_waiting_guard timeout is enough for
+    // transformMPI to post the IRecv.
+    std::atomic_bool pg_guard = false;
+    auto after_pg = chain() | ex::then([&pg_guard](auto&& pg) {
+                      pg_guard = true;
+                      return std::move(pg);
+                    }) |
+                    ex::ensure_started();
+    try_waiting_guard(pg_guard);
+    EXPECT_TRUE(pg_guard);
+
+    // "ensure" (by waiting a reasonable amount of time) that IRecv doesn't complete
+    std::this_thread::sleep_for(std::chrono::milliseconds(200));
+    EXPECT_FALSE(sent_guard);
+
+    // Note:
+    // At this point we checked if everything went as expected in the posting phase. Let's signal
+    // Rank 1 that it can send back the message that will unlock IRecv.
+    const int signal = 13;
+    MPI_Send(&signal, 1, MPI_INT, 1, 0, world);
```

Comment:
I'm not sure I get this part. The `MPI_Send` is in standard mode, so it's implementation dependent whether the call will complete before the matching `MPI_Recv` is issued. The call may complete as soon as the data is buffered (but not yet sent). Perhaps use `MPI_Rsend`? this will start the `send` only when the matching `recv` is posted.

Reply:
Not sure I understand the question. This MPI_Send is just triggering the other rank to send the message that match the receive posted with trasformMPI. When it returns doesn't influence the test.

The timeline in the following:
```
RANK 0:                                            |  Rank 1:
- Post MPI_Irecv via tranformMPI (l. 67)           | - MPI_Recv (l. 123)
- check that the comm pipeline has unlocked (92)   |    |
  (as the MPI_Irecv can be immediately posted)     |    |
- check that the continuation of the transformMPI  |    |
  hasn't been unlocked (l. 96)                     |    |
- MPI_Send (l. 102)        ------------------------|--> Ok
(Match with previous Irecv) <----------------------| - MPI_Send (l. 129)
- check that the continuation is now unlocked (111)
```
Note that all checks are performed with a delay, as certain operations might not execute immediately.